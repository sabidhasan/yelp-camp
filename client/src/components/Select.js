import React from 'react'

class Select extends React.Component {
  constructor() {
    super();

    this.state = {
      expanded: false,
      value: null
    }
    this.toggleExpanded = this.toggleExpanded.bind(this);
  }

  updateValue(newValueObject) {
    if (newValueObject === null) {
      // Reset the state
      this.setState({value: null})
    } else {
      this.setState({value: newValueObject.text})
    }
    if (this.props.onChange) {
      this.props.onChange(newValueObject ? newValueObject.value : null)
    }
  }

  toggleExpanded() {
    this.setState({expanded: !this.state.expanded})
  }

  render() {
    const items = this.props.items || []
    return (
      <div
        tabindex='0'
        className={`Select${this.state.expanded ? ' Select--expanded' : ''}`}
        onClick={this.toggleExpanded}>
        <span
          className={`Select__current-value bold ${this.state.expanded ? 'Select__current-value--active' : ''}`}>
          {this.state.value || this.props.defaultText}
        </span>
        <ul className={`Select__items ${this.state.expanded ? 'Select__items--expanded' : ''}`}>
          <li
            className='Select__item'
            onClick={()=> this.updateValue(null)}
            key={-1}>
            {this.props.defaultText}
          </li>
          {items
            .map((v, i) => (
              <li
                className='Select__item'
                onClick={() => this.updateValue(v)}
                key={i}
                aria-setsize={items.length}
                aria-posinset={i + 1}
                >
                {v.text}
              </li>
            ))
          }
        </ul>
    </div>
  )}
}

export default Select
