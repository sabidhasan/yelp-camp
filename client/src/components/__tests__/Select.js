import React from 'react'
import toJson from 'enzyme-to-json'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import Select from '../Select'

configure({ adapter: new Adapter() });

describe('<Select />', () => {
  let component;
  let props;
  let onChange;

  beforeEach(() => {
    onChange = jest.fn();
    props = {
      items: [
        {value: 1, text: 'First'},
        {value: 2, text: 'Second'},
        {value: 3, text: 'Third'}
      ],
      defaultText: 'Test Text',
      onChange
    }
    component = shallow(<Select { ...props } />)
  })

  it('Component should be defined', () => {
    expect(component).toBeDefined();
  });

  it('Control is unexpanded upon load', () => {
    expect(component.find('.Select--expanded')).toHaveLength(0);
  });

  it('Value presented to user is default text upon load', () => {
    expect(component.find('.Select__current-value').text()).toBe('Test Text');
  });

  it('All items are presented to the user', () => {
    expect(component.find('.Select__item--main')).toHaveLength(3);
  });

  it('Clicking already-selected item does not change text', () => {
    component.find('.Select__item--first').simulate('click');
    expect(component.find('.Select__current-value').text()).toBe('Test Text');
  });

  it('Clicking already-selected item does not call callback function', () => {
    component.find('.Select__item--first').simulate('click');
    expect(onChange).toHaveBeenCalledTimes(0);
  });

  it('Clicking a new item changes the displayed text', () => {
    component.find('.Select__item--main').first().simulate('click');
    expect(component.find('.Select__current-value').text()).toBe('First');
  });

  it('Clicking a new item calls callback function with correct argument', () => {
    component.find('.Select__item--main').first().simulate('click');
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(onChange).toBeCalledWith(1);
  });

  it('Select component snapshot matches what is expected', () => {
    expect(toJson(component)).toMatchSnapshot();
  })
})
