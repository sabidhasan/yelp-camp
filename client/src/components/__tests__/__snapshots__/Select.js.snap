// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> Select component snapshot matches what is expected 1`] = `
<div
  className="Select"
  onClick={[Function]}
  tabIndex="0"
>
  <span
    className="Select__current-value bold "
  >
    Test Text
  </span>
  <ul
    className="Select__items "
  >
    <li
      className="Select__item Select__item--first"
      key="-1"
      onClick={[Function]}
    >
      Test Text
    </li>
    <li
      aria-posinset={1}
      aria-setsize={3}
      className="Select__item Select__item--main"
      key="0"
      onClick={[Function]}
    >
      First
    </li>
    <li
      aria-posinset={2}
      aria-setsize={3}
      className="Select__item Select__item--main"
      key="1"
      onClick={[Function]}
    >
      Second
    </li>
    <li
      aria-posinset={3}
      aria-setsize={3}
      className="Select__item Select__item--main"
      key="2"
      onClick={[Function]}
    >
      Third
    </li>
  </ul>
</div>
`;
